### PR TITLE
feature: discover migrations

### PR DESCRIPTION
**Pull request by joelbutcher**
## TL;DR

This PR allows developers to specify whether their package should auto-discover all migrations in their package without having to specify individual file names.

## Context – What problem does this solve?

For smaller packages with maybe a handful of migrations, calling `hasMigration` multiple times (or just `hasMigrations` once) is trivial and non-invasive.

However for larger packages, or for people using this tool in larger projects (e.g. registering composer modules in a large-scale production application), having to specify the name of every migration that is within a package or module becomes arduous and prone to mistakes

For example, newer developers will not necessarily always remember that each migration **must** be manually registered in a service provider (especially since Laravel's functionality is to auto-discover migrations).

## Solution

This PR allows developers to choose whether or not their package "auto-discovers" its own migrations, by expanding the `Package` API with a new `discoversMigrations` method. 

Developers and teams can now simply build a new package with as many migrations as they like without needing to remember to manually register the migration:

```php
$package
    ->name('your-package-name')
    ->discoversMigrations();
```

When this is called, we will assume this is the default and that any calls to `hasMigration` or `hasMigrations` can safely be ignored. (If not, then in some scenarios, this would cause duplicate migrations and errors after publishing migrations and running `artisan migrate`).

```php
$package
    ->name('your-package-name')
    ->hasMigration('create_my_package_tables') // Ignored
    ->hasMigrations(['my_package_tables', 'some_other_migration'])  // Ignored
    ->discoversMigrations();
```

I've also updated the `hasMigrations` to call `discoversMigrations` when no migration names are provided

### CODE REVIEW

**Comment by erikgaal** [comment]
Maybe add a (default) parameter of `database/migrations` so that you could technically discover it from elsewhere.

File: src/PackageServiceProvider.php [hunk 1]
```diff
@@ -213,4 +198,56 @@ public function packageView(?string $namespace)
...
+                $this->loadMigrationsFrom($filePath);
+            }
+        }
+    }
+
+    protected function discoverMigrations(): void
+    {
+        $now = Carbon::now();
+
+        $files = (new Filesystem())->files($this->package->basePath('/../database/migrations'));
```
Comment (freekmurze) [inline src/PackageServiceProvider.php:235]: What happens if this path does not exist?

File: src/Package.php [hunk 2]
```diff
@@ -168,6 +177,12 @@ public function hasMigration(string $migrationFileName): static
 
     public function hasMigrations(...$migrationFileNames): static
     {
+        if (empty($migrationFileNames)) {
```
Comment (freekmurze) [inline src/Package.php:180]: So for packages that don't have migrations explicitly specified, we'd auto discover them by default.

This will cause that for all packages that don't have migrations at all, we'll still check the `migrations` directory, which has an - albeit small - performance impact.

I'd be more comfortable merging this in, if people would only specifically opt-in to the checking behavior. What do you think?

Comment (joelbutcher) [reply src/Package.php:180]: @freekmurze you think remove this check and just require package devs to explicitly say "just discover my migrations"

File: src/PackageServiceProvider.php (hunk 1 above)
Comment (joelbutcher) [reply src/PackageServiceProvider.php:235]: Symfony throws a directory not found exception.

I'll allow package developers to define a `$path` parameter for `discoversMigrations` and default it to `'/../database/migrations'` as @erikgaal

**Comment by freekmurze** [comment]
Thanks!
